src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -0,0 +1,47 @@
+/**
+ * @file bootstrap.hpp
+ * @author Ryan Curtin
+ *
+ * Implementation of the Bootstrap() function, which creates a bootstrapped
+ * dataset from the given input dataset.
+ */
+#ifndef MLPACK_METHODS_RANDOM_FOREST_BOOTSTRAP_HPP
+#define MLPACK_METHODS_RANDOM_FOREST_BOOTSTRAP_HPP
+
+namespace mlpack {
+namespace tree {
+
+/**
+ * Given a dataset, create another dataset via bootstrap sampling, with labels.
+ */
+template<bool UseWeights,
+         typename MatType,
+         typename LabelsType,
+         typename WeightsType>
+void Bootstrap(const MatType& dataset,
+               const LabelsType& labels,
+               const WeightsType& weights,
+               MatType& bootstrapDataset,
+               LabelsType& bootstrapLabels,
+               WeightsType& bootstrapWeights)
+{
+  bootstrapDataset.set_size(dataset.n_rows, dataset.n_cols);
+  bootstrapLabels.set_size(labels.n_elem);
+  if (UseWeights)
+    bootstrapWeights.set_size(weights.n_elem);
+
+  // Random sampling with replacement.
+  for (size_t i = 0; i < dataset.n_cols; ++i)
+  {
+    const size_t index = math::RandInt(dataset.n_cols);
```

Comment:
A simple test:

```
#include <mlpack/core.hpp>
#include <mlpack/methods/random_forest/bootstrap.hpp>

using namespace mlpack;
using namespace arma;

int main()
{
  mat dataset(10, 100000, arma::fill::randu);
  Col<size_t> labels(100000, fill::zeros);
  vec weights;

  mat newset;
  Col<size_t> newlabels;

  for (size_t i = 0; i < 1000; ++i)
  {
    tree::Bootstrap<false>(dataset, labels, weights, newset, newlabels, weights);
  }
}
```

With randi:

```
$ time ./test-randi

real    0m20.356s
user    0m20.376s
sys     0m0.348s
```

Without (i.e. the old code):

```
$ time ./test

real    0m33.492s
user    0m33.512s
sys     0m0.348s
```

I had not thought this could be slow!  Thanks for pointing it out.